// @vtex/api/lib/service/tracing/metrics/measurers/EventLoopLagMeasurer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const perf_hooks_1 = require("perf_hooks");
const utils_1 = require("../../../../utils");
const instruments_1 = require("../instruments");
class EventLoopLagMeasurer {
    constructor() {
        this.eventLoopDelayMonitor = perf_hooks_1.monitorEventLoopDelay({ resolution: 10 });
        this.percentilesInstrument = instruments_1.createEventLoopLagPercentilesInstrument();
        this.maxInstrument = instruments_1.createEventLoopLagMaxInstrument();
    }
    start() {
        this.eventLoopDelayMonitor.enable();
    }
    async updateInstrumentsAndReset() {
        this.maxInstrument.set(utils_1.nanosecondsToSeconds(this.eventLoopDelayMonitor.max));
        this.setPercentileObservation(95);
        this.setPercentileObservation(99);
        this.eventLoopDelayMonitor.reset();
    }
    setPercentileObservation(percentile) {
        this.percentilesInstrument.set({ ["percentile" /* PERCENTILE */]: percentile }, utils_1.nanosecondsToSeconds(this.eventLoopDelayMonitor.percentile(percentile)));
    }
}
exports.EventLoopLagMeasurer = EventLoopLagMeasurer;
