// @vtex/intl-equalizer/dist/missingKeysTable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createTable;

var _cliTable = require('cli-table3');

var _cliTable2 = _interopRequireDefault(_cliTable);

var _safe = require('colors/safe');

var _safe2 = _interopRequireDefault(_safe);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createTable(result, referenceLocale) {
  const table = new _cliTable2.default({
    chars: {
      top: '═',
      'top-mid': '╤',
      'top-left': '╔',
      'top-right': '╗',
      bottom: '═',
      'bottom-mid': '╧',
      'bottom-left': '╚',
      'bottom-right': '╝',
      left: '║',
      'left-mid': '╟',
      mid: '─',
      'mid-mid': '┼',
      right: '║',
      'right-mid': '╢',
      middle: '│'
    },
    head: [{
      colSpan: 3,
      hAlign: 'center',
      content: _safe2.default.yellow(`MISSING KEYS`)
    }]
  });

  table.push([{
    colSpan: 3,
    hAlign: 'center',
    content: _safe2.default.yellow(`REFERENCE LOCALE: ${referenceLocale.toUpperCase()}`)
  }]);

  table.push([{ hAlign: 'center', content: _safe2.default.yellow('LOCALE') }, { hAlign: 'center', content: _safe2.default.yellow('PATH') }, { hAlign: 'center', content: _safe2.default.yellow('MISSING KEYS') }]);

  Object.keys(result).forEach(countryName => {
    if (countryName === referenceLocale) return;

    const country = countryName.toUpperCase();
    const filepath = result[countryName].filepath;
    const missingKeys = result[countryName].missingKeys.reduce((acc, curr) => acc + curr.key + '\n', '');
    table.push([country, filepath, missingKeys]);
  });

  console.log(table.toString());
}
